components/raftstore/src/store/fsm/sync_policy.rs

```diff
@@ -0,0 +1,322 @@
+    /// to sync, then this thread should not sync.
+    fn check_sync_internal(&mut self, before_sync_ts: i64) -> bool {
+        let last_sync_ts = self.global_last_sync_ts.load(Ordering::Acquire);
+        let plan_sync_ts = self.global_plan_sync_ts.load(Ordering::Acquire);
+        if last_sync_ts != plan_sync_ts {
+            // Another thread is planning to sync, so this thread should do nothing
+            assert!(
+                plan_sync_ts > last_sync_ts,
+                "plan sync ts {} < last sync ts {}",
+                plan_sync_ts,
+                last_sync_ts
+            );
+            return false;
+        }
+        if before_sync_ts <= last_sync_ts {
+            return false;
+        }
+
+        let mut need_sync = if self.unsynced_regions_map.len() > UNSYNCED_REGIONS_SIZE_LIMIT {
+            self.metrics.sync_events.sync_raftdb_delay_cache_is_full += 1;
+            true
+        } else {
+            false
+        };
+
+        need_sync |= {
+            let elapsed = before_sync_ts - last_sync_ts;
+            self.metrics
+                .thread_check_result
+                .observe(elapsed as f64 / 1e9);
+
+            if elapsed >= self.delay_sync_us {
+                self.metrics.sync_events.sync_raftdb_reach_deadline += 1;
+                true
+            } else {
+                false
+            }
+        };
+
+        if !need_sync {
+            return false;
+        }
+
+        // If it's false, it means another thread is planning to sync, so this thread should do nothing
+        plan_sync_ts
+            == self.global_plan_sync_ts.compare_and_swap(
+                plan_sync_ts,
+                before_sync_ts,
+                Ordering::AcqRel,
+            )
+    }
+
+    /// Return if all unsynced region are flushed
+    fn flush_unsynced_regions(&mut self, synced_ts: i64, flush_all: bool) -> bool {
+        while let Some((ts, region_id)) = self.unsynced_regions_set.iter().next() {
+            let delay_duration = synced_ts - *ts;
+            if !flush_all && delay_duration <= 0 {
+                break;
+            }
+            self.metrics
+                .sync_delay_duration
+                .observe(delay_duration as f64 / 1e9);
+
+            let (ts, region_id) = (*ts, *region_id);
+            assert_eq!(self.unsynced_regions_set.remove(&(ts, region_id)), true);
+            let (number, notifier, ts2) = self.unsynced_regions_map.remove(&region_id).unwrap();
+            assert_eq!(ts, ts2);
+
+            loop {
+                let pre_number = notifier.load(Ordering::Acquire);
+                assert_ne!(pre_number, number);
+                if pre_number > number {
+                    break;
+                }
+                if pre_number == notifier.compare_and_swap(pre_number, number, Ordering::AcqRel) {
+                    if let Err(e) = self.router.force_send(region_id, PeerMsg::Noop) {
```

Comment:
Could we send the number of current persisted ready so that avoid hold this atomic variable?

Reply:
The previous code is like what you said. The atomic variable is used to speed up the notify process and the no-op msg is just used to wake up the peer fsm if it's in sleep.